### PR TITLE
Optimized gas a bit

### PR DESCRIPTION
**Pull request by k06a**
Added some transactions to tests to enable `eth-gas-reporter` to collect something.

Before:
```
····················|·························|·············|·············|·············|··············|
|  Contract         ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls     ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseInputTx           ·      37515  ·      41205  ·      39886  ·           5  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseOutputTx          ·      29471  ·      30602  ·      30234  ·           9  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  validateHeaderChainTx  ·      55408  ·     191309  ·     114373  ·           7  ·
····················|·························|·············|·············|·············|··············|
```
After:
```
····················|·························|·············|·············|·············|··············|
|  Contract         ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls     ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseInputTx           ·      37248  ·      40667  ·      39402  ·           5  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseOutputTx          ·      28934  ·      30065  ·      29726  ·           9  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  validateHeaderChainTx  ·      55408  ·     183537  ·     110563  ·           7  ·
····················|·························|·············|·············|·············|··············|
```

### CODE REVIEW

**Comment by k06a** [comment]
Tried to optimize `slice()` method:
```
····················|·························|·············|·············|·············|··············|
|  Contract         ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls     ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseInputTx           ·      37038  ·      40385  ·      39135  ·           5  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseOutputTx          ·      28865  ·      29927  ·      29578  ·           9  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  validateHeaderChainTx  ·      55408  ·     182154  ·     109892  ·           7  ·
····················|·························|·············|·············|·············|··············|
```

Have to check if this optimization works for these tests only or for any length.

**Comment by ErinHales** [comment]
Let's implement the `keccak256Slice` function in `BTCUtils.sol` as well.

**Comment by ErinHales** [comment]
> Let's implement the `keccak256Slice` function in `BTCUtils.sol` as well.

I think you might have misunderstood me. `keccak256Slice` is fine to stay in `BytesLib`.  I meant you can just implement it in `BTCUtils` like you did in `ValidateSPV`. 
Example(line 151): 
```
    function isLegacyInput(bytes memory _input) internal pure returns (bool) {
        return keccak256(_input.slice(36, 1)) != keccak256(hex"00");
    }
```
should be:
```
    function isLegacyInput(bytes memory _input) internal pure returns (bool) {
        return _input.keccak256Slice(36, 1)) != keccak256(hex"00");
    }
```

**Comment by k06a** [comment]
@ErinHales got it, but for some reason, I don't see any gas improvements 🤷‍♂️
```
····················|·························|·············|·············|·············|
|  Contract         ·  Method                 ·  Min        ·  Max        ·  Avg        ·
····················|·························|·············|·············|·············|
|  ValidateSPVTest  ·  parseInputTx           ·      37047  ·      40397  ·      39146  ·
····················|·························|·············|·············|·············|
|  ValidateSPVTest  ·  parseOutputTx          ·      28868  ·      29933  ·      29562  ·
····················|·························|·············|·············|·············|
|  ValidateSPVTest  ·  validateHeaderChainTx  ·      55408  ·     182223  ·     109925  ·
····················|·························|·············|·············|·············|
```

**Comment by k06a** [comment]
Achieved 2x optimization for `validateHeaderChain`:
```
····················|·························|·············|·············|·············|··············|
|  Contract         ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls     ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseInputTx           ·      37047  ·      40397  ·      39146  ·           5  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  parseOutputTx          ·      28868  ·      29933  ·      29562  ·           9  ·
····················|·························|·············|·············|·············|··············|
|  ValidateSPVTest  ·  validateHeaderChainTx  ·      52585  ·      90622  ·      65061  ·           7  ·
····················|·························|·············|·············|·············|··············|
```

**Comment by ErinHales** [comment]
> @ErinHales got it, but for some reason, I don't see any gas improvements 🤷‍♂️
> 
> ```
> ····················|·························|·············|·············|·············|··············|
> |  Contract         ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls     ·
> ····················|·························|·············|·············|·············|··············|
> |  ValidateSPVTest  ·  parseInputTx           ·      37047  ·      40397  ·      39146  ·           5  ·
> ····················|·························|·············|·············|·············|··············|
> |  ValidateSPVTest  ·  parseOutputTx          ·      28868  ·      29933  ·      29562  ·           9  ·
> ····················|·························|·············|·············|·············|··············|
> |  ValidateSPVTest  ·  validateHeaderChainTx  ·      55408  ·     182223  ·     109925  ·           7  ·
> ····················|·························|·············|·············|·············|··············|
> ```

Awesome, thanks!  It's probably because these tests don't cover the changes made in `BTCUtils`

**Comment by prestwich** [comment]
So this might be a solidity compiler artifact. Generally, library functions that are marked `internal` get compiled into the calling contract. This means that you won't see gas savings by duplicating the function across multiple files

**Comment by k06a** [comment]
@prestwitch I first though @ErinHales asked me to relocate this method from one lib to another. Then I realized he asked to use `keccak256Slice` in more places. And, yep it seems these tests do not cover BTCUtil code 👌

File: solidity/contracts/BTCUtils.sol [hunk 1]
```diff
@@ -52,6 +52,29 @@ library BTCUtils {
         return _newValue;
     }
 
+    /// @notice          Changes the endianness of a uin256
```
Comment (prestwich) [inline solidity/contracts/BTCUtils.sol]: typo: `uin` => `uint`

**Comment by prestwich** [comment]
I'm going to give this more review this week, then merge and tag a release over the weekend provided nothing serious crops up

**Pull request merged**